File: Week_1/HappyNumber.java
import java.util.HashSet;
import java.util.Scanner;
import java.util.Set;

public class HappyNumber {
    public static boolean isHappy(int n) {
        Set<Integer> seen = new HashSet<>();
        
        while (n != 1 && !seen.contains(n)) {
            seen.add(n);
            n = sumOfSquares(n);
        }
        
        return n == 1;
    }

    private static int sumOfSquares(int n) {
        int sum = 0;
        while (n > 0) {
            int digit = n % 10;
            sum += digit * digit;
            n /= 10;
        }
        return sum;
    }

    public static void main(String[] args) {
        // Using try-with-resources to handle the scanner
        try (Scanner scanner = new Scanner(System.in)) {
            System.out.print("Enter a number: ");
            int number = scanner.nextInt();
            if (isHappy(number)) {
                System.out.println(number + " is a happy number!");
            } else {
                System.out.println(number + " is not a happy number.");
            }
        } 
    }
}

File: Week_1/HappyNumber.py
def is_happy(n: int):
    s = set()

    while n != 1 and n not in s:
        s.add(n)
        n = sum_of_squares(n)

    return n

def sum_of_squares(n: int):
    total = 0
    while n > 0:
        digit = n % 10
        total += digit * digit
        n //= 10
    return total

number = int(input("Enter a number: "))
result = is_happy(number)

if result == 1:
    print(f"{number} is a happy number!")
else:
    print(f"{number} is not a happy number.")
